src/main/flight/pos_hold.c

```diff
@@ -0,0 +1,112 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Betaflight is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Betaflight. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "platform.h"
+
+#ifdef USE_POS_HOLD_MODE
+
+#include "math.h"
+#include "build/debug.h"
+#include "common/maths.h"
+
+#include "config/config.h"
+#include "fc/runtime_config.h"
+#include "fc/rc.h"
+#include "flight/failsafe.h"
+#include "flight/position.h"
+#include "flight/autopilot.h"
+#include "rx/rx.h"
+
+#include "pos_hold.h"
+
+posHoldState_t posHold;
+
+void posHoldReset(void)
+{
+    posHold.posHoldIsOK = true;
+    posHold.targetLocation = gpsSol.llh;
+    resetPositionControl(posHold.targetLocation);
+}
+
+void posHoldInit(void)
+{
+    posHold.isPosHoldRequested = false;
+    posHold.deadband = rcControlsConfig()->pos_hold_deadband / 100.0f;
+    posHold.useStickAdjustment = rcControlsConfig()->pos_hold_deadband;
+}
+
+void posHoldProcessTransitions(void)
+{
+    if (FLIGHT_MODE(POS_HOLD_MODE)) {
+        if (!posHold.isPosHoldRequested) {
+            // initialise relevant values each time position hold starts
+            posHoldReset();
+            posHold.isPosHoldRequested = true;
+        }
+    } else {
+        posHold.isPosHoldRequested = false;
+    }
+}
+
+void posHoldUpdateTargetLocation(void)
+{
+    // if failsafe is active, eg landing mode, don't update the original start point 
+    if (!failsafeIsActive()) {
+        // otherwise if sticks are not centered, allow start point to be updated
+        if (posHold.useStickAdjustment) {
+            if ((getRcDeflectionAbs(FD_ROLL) > posHold.deadband) || (getRcDeflectionAbs(FD_PITCH) > posHold.deadband))  {
+                // allow user to fly the quad, in angle mode, when sticks are outside the deadband
+                // while sticks are outside the deadband,
+                // keep updating the home location to the current GPS location each iteration
+                posHoldReset();
+            }
+        }
+    }
+}
+
+void posHoldUpdate(void)
+{
+    posHoldUpdateTargetLocation();
+
+    if (getIsNewDataForPosHold() && posHold.posHoldIsOK) {
+        posHold.posHoldIsOK = positionControl(posHold.deadband);
+    }
+}
+
+void updatePosHoldState(timeUs_t currentTimeUs) {
+    UNUSED(currentTimeUs); 
+
+    // check for enabling Alt Hold, otherwise do as little as possible while inactive
+    posHoldProcessTransitions();
+
+    if (posHold.isPosHoldRequested) {
+        posHoldUpdate();
+    } else {
+        posHoldAngle[AI_PITCH] = 0.0f;
+        posHoldAngle[AI_ROLL] = 0.0f;
+    }
+}
+
+bool showPosHoldWarning(void) {
```

Comment:
Maybe `posHoldFailure` since it applies whenever the user flips the switch, and either position hold fails to start, or fails due to a sanity check.

Reply:
Then maybe `posHoldIsFailing()` is a better name. Just depends from which side you look at it (inversion of boolean logic).

With my name suggestion I was implying that you need to invert the logic but I didn't explicitly say it :)

Reply:
I will stick with `posHoldFailure` because that's exactly what it is; position hold was requested by sticks, but either it failed to start, or failed after starting.  It's not that it is in the process of failing; it has failed.
I've refactored this a little bit just now, but it works the same.